Expand Static Constructor C# documentation (#12733)
dotnet/docs: docs/csharp/programming-guide/classes-and-structs/static-constructors.md

 
## Remarks
Static constructors have the following properties:  
  
- A static constructor does not take access modifiers or have parameters.  

- A class or struct can only have one static constructor.

- Static constructors cannot be inherited or overloaded.

- A static constructor cannot be called directly and is only meant to be called by the common language runtime (CLR). It is invoked automatically.

- The user has no control on when the static constructor is executed in the program.
  
- A static constructor is called automatically to initialize the [class](../../../csharp/language-reference/keywords/class.md) before the first instance is created or any static members are referenced. A static constructor will run before an instance constructor. Note that a type's static constructor is called when a static method assigned to an event or a delegate is invoked and not when it is assigned. If static field variable initializers are present in the class of the static constructor, they will be executed in the textual order in which they appear in the class declaration immediately prior to the execution of the static constructor.

> - If you don't provide a static constructor to initialize static fields, the C# compiler will supply a default implicit static constructor that initializes static fields to their default value as listed in the [Default Values Table](../../../csharp/language-reference/keywords/default-values-table.md). 

Reviewer: I'd like to rephrase this. The end result (initializing values to their default) is correct, but the compiler is not required to supply a static constructor. See the [last paragraph on default values in the C# language spec](https://docs.microsoft.com/dotnet/csharp/language-reference/language-specification/variables#default-values).

```suggestion
- If you don't provide a static constructor to initialize static fields, all static fields are initialized to their default value as listed in the [Default Values Table](../../../csharp/language-reference/keywords/default-values-table.md). 
```

Author: Appreciate the input.

That line was incorporated from [Constructors (C# Programming Guide)](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/constructors#static-constructors) :
>...If you don't provide a static constructor to initialize static fields, the C# compiler will supply a default static constructor that initializes static fields to their default value as listed in the Default Values Table.

Also, similar phrasing is used in [Static constructors and the TypeInitializationException](https://docs.microsoft.com/en-us/dotnet/api/system.typeinitializationexception?view=netframework-4.8#static-constructors-and-the-typeinitializationexception):
> ...Static constructors can be explicitly defined by a developer. If a static constructor is not explicitly defined, compilers automatically create one to initialize any static (in C#) or Shared (in Visual Basic) members of the type.

Should the text on these other pages be revised?

Reviewer: 😢 Yes, those pages should be updated as well. We're moving toward ensuring that the articles in these sections use language that's consistent with the spec (although often more informal). These should agree with the spec, and not imply different behavior.

Do you want to make those additional changes, or should I create an issue to update those docs and reference this conversation?

Author: I'm not sure how to best rephrase the text, so it'd be better if you were to file the issue.

Also, having looked at the IL output on [SharpLab](https://sharplab.io), I'm not sure the phrasing is incorrect.  When a static field is initialized at declaration, a `.cctor` is generated within which the static fields are initialized.

```csharp
//no .cctor
public class A{
    public static int x;
}
```
```csharp
//has .cctor 
//initializiation takes place in .cctor
public class B{
    public static int x=1;
}
```
```csharp
//has .cctor 
//initializiation takes place in .cctor
//x initialized before  y
public class C{
    public static int x=1;
    static C(){int y=2;}
}
```

- The presence of a static constructor prevents the addition of the the <xref:System.Reflection.TypeAttributes.BeforeFieldInit> type attribute. This limits runtime optimization.

- A field declared as `static readonly` may only be assigned as part of its declaration or in a static constructor. When an explicit static constructor is not required, initialize static fields at declaration, rather than through a static constructor for better runtime optimization.

> [!Note]
> Though not directly accessible, the presence of an explicit static constructor should be documented to assist with troubleshooting initialization exceptions.

### Usage

- A typical use of static constructors is when the class is using a log file and the constructor is used to write entries to this file.  
- Static constructors are also useful when creating wrapper classes for unmanaged code, when the constructor can call the `LoadLibrary` method.  

- Static constructors are also a convenient place to enforce run-time checks on the type parameter that cannot be checked at compile-time via constraints (Type parameter constraints).

### Explicit initialization

> > [!CAUTION]

Reviewer: I think that leaving this section out would be better. Readers that already know about `RunClassConstructor` would (hopefully) know they should not call the method. Other readers can remain blissfully unaware of its existence.

What were your thoughts regarding adding this?

Author: The `RunClassConstructor` was mainly added for completion's sake . It has been removed as its presence would promote misuse.

It may be worthwhile to include an example that an empty static void method would serve the same purpose.

```csharp
using System; 

public class B 
{      
    public static void TriggerStaticConstructor() {} 
	
    static B(){
        Console.WriteLine("Static constructor invoked.");
    }
} 
public class A 
{
    public static void Main(string[] args){
        B.TriggerStaticConstructor();
    }
} 
```

Reviewer: Thanks for the update. I think this addition isn't necessary, so I'll :shipit: now.

> [!Note]
> Though not directly accessible, the presence of an explicit static constructor should be documented to assist with troubleshooting initialization exceptions.

## Usage

- A typical use of static constructors is when the class is using a log file and the constructor is used to write entries to this file.  
- Static constructors are also useful when creating wrapper classes for unmanaged code, when the constructor can call the `LoadLibrary` method.  
  
  
## Example  
 In this example, class `Bus` has a static constructor. When the first instance of `Bus` is created (`bus1`), the static constructor is invoked to initialize the class. The sample output verifies that the static constructor runs only one time, even though two instances of `Bus` are created, and that it runs before the instance constructor runs.  
  
 [!code-csharp[csProgGuideObjects#15](~/samples/snippets/csharp/VS_Snippets_VBCSharp/csProgGuideObjects/CS/Objects.cs#15)]
 
## C# language specification  

> For more information, see the [Static constructors](~/_csharplang/spec/classes.md#static-constructors) section of the [C# language specification](../../language-reference/language-specification/index.md).

Reviewer 2: There is no *language-specification* folder in this repo anymore and the referenced index page is redirected to the one in this suggestion from the csharplang repo

```suggestion
For more information, see the [Static constructors](~/_csharplang/spec/classes.md#static-constructors) section of the [C# language specification](~/_csharplang/spec/introduction.md).
```
